File: extensions/positron-sql/src/cursor.ts
/*---------------------------------------------------------------------------------------------
 *  Copyright (C) 2026 Posit Software, PBC. All rights reserved.
 *  Licensed under the Elastic License 2.0. See LICENSE.txt for license information.
 *--------------------------------------------------------------------------------------------*/

/**
 * Reading what the text immediately before the cursor says about what is being completed.
 *
 * Deliberately a scan of the raw text rather than anything the parser produced. Completion is
 * asked for while the user is midway through typing an identifier, which is the one state the
 * tokenizer cannot always be run on: a half typed `"Order T` has no closing quote, so it would
 * swallow the rest of the document. The few characters before the cursor say everything that is
 * needed -- the partial name, the dotted qualifiers in front of it, and whether the cursor is
 * somewhere no identifier belongs at all.
 */

/** Characters that may appear in a bare identifier. */
const IDENTIFIER_CHAR = /[A-Za-z0-9_$]/;

/** Closing quote to opening quote, for reading a quoted identifier backwards from the cursor. */
const CLOSING_QUOTES = new Map([['"', '"'], ['`', '`'], [']', '[']]);

/**
 * Opening quote to closing quote, for the identifier forms the common dialects use: standard SQL
 * and PostgreSQL double quotes, MySQL backticks, T-SQL brackets. Accepting all three whatever the
 * dialect costs nothing here -- a dialect that does not use a form will not have produced one.
 */
const IDENTIFIER_QUOTES = new Map([['"', '"'], ['`', '`'], ['[', ']']]);

/** Stands in for the identifier being typed, so the text around it can be tokenized. */
export const PLACEHOLDER = '__positron_sql_cursor__';

export interface Cursor {
	/** The partial identifier under the cursor, possibly empty. */
	readonly prefix: string;

	/** Where the partial identifier starts, and so what a matching completion is filtered by. */
	readonly prefixStart: number;

	/**
	 * Where a completion's edit should start.
	 *
	 * The same as {@link prefixStart} except inside a half typed quoted identifier, where it is
	 * the opening quote. A completion that needs quoting inserts them itself, so an edit starting
	 * after the quote would turn `o."Order T` into `o.""Order Total"`.
	 */
	readonly replaceStart: number;

	/**
	 * The dotted names in front of the prefix, outermost first: `sales.orders.` gives
	 * `['sales', 'orders']`. Empty when the cursor is not after a dot.
	 */
	readonly qualifiers: readonly string[];

	/**
	 * Set when the cursor is somewhere no identifier belongs -- inside a string literal or a line
	 * comment -- and the request should produce nothing at all.
	 */
	readonly suppressed: boolean;

	/** Whether the partial identifier began with an opening quote. */
	readonly quoted: boolean;
}

/**
 * Finds the quote or comment the cursor sits inside, if one was opened on the cursor's line.
 *
 * Returns the opening delimiter and the offset just past it, or undefined when the cursor is in
 * ordinary code. Only the cursor's own line is scanned: an unterminated literal is by definition
 * something the tokenizer cannot get past, so there is no token stream to consult, and a
 * delimiter opened on an earlier line is far rarer than the case this exists for -- a user part
 * way through typing `"Order Total"`.
 */
function openQuoteAt(text: string, offset: number): { delimiter: string; contentStart: number } | undefined {
	let position = text.lastIndexOf('\n', Math.max(0, offset - 1)) + 1;
	if (offset === 0) {
		position = 0;
	}

	while (position < offset) {
		const char = text[position];

		if (char === '-' && text.startsWith('--', position)) {
			return { delimiter: '--', contentStart: position + 2 };
		}

		if (char === '\'') {
			position += 1;
			let closed = false;
			while (position < offset) {
				if (text[position] === '\'') {
					// A doubled quote is an escaped one and does not close the literal.
					if (text.startsWith('\'\'', position)) {
						position += 2;
						continue;
					}
					closed = true;
					break;
				}
				position += 1;
			}
			if (!closed) {
				return { delimiter: '\'', contentStart: position };
			}
			position += 1;
			continue;
		}

		const closing = IDENTIFIER_QUOTES.get(char);
		if (closing !== undefined) {
			const end = text.indexOf(closing, position + 1);
			if (end < 0 || end >= offset) {
				return { delimiter: char, contentStart: position + 1 };
			}
			position = end + 1;
			continue;
		}

		position += 1;
	}

	return undefined;
}

/**
 * Reads the identifier ending at `end`, returning its text and start offset.
 *
 * Handles the quoted forms as well as bare ones, so `"Order Details".` is recognized as a
 * qualifier rather than read as the bare word `Details`.
 */
function readIdentifierBackwards(text: string, end: number): { name: string; start: number } | undefined {
	if (end <= 0) {
		return undefined;
	}

	const opening = CLOSING_QUOTES.get(text[end - 1]);
	if (opening !== undefined) {
		const start = text.lastIndexOf(opening, end - 2);
		if (start < 0) {
			return undefined;
		}
		return { name: text.slice(start + 1, end - 1), start };
	}

	let start = end;
	while (start > 0 && IDENTIFIER_CHAR.test(text[start - 1])) {
		start -= 1;
	}
	return start === end ? undefined : { name: text.slice(start, end), start };
}

/** Reads the partial identifier at the cursor and any dotted qualifiers in front of it. */
export function readCursor(text: string, offset: number): Cursor {
	const openQuote = openQuoteAt(text, offset);
	if (openQuote && (openQuote.delimiter === '\'' || openQuote.delimiter === '--')) {
		return {
			prefix: '',
			prefixStart: offset,
			replaceStart: offset,
			qualifiers: [],
			suppressed: true,
			quoted: false,
		};
	}

	let prefixStart: number;
	let quoteStart: number;
	if (openQuote) {
		// Part way through a quoted identifier: everything since the quote is the prefix, and the
		// qualifier scan continues from the quote itself.
		prefixStart = openQuote.contentStart;
		quoteStart = prefixStart - 1;
	} else {
		prefixStart = offset;
		while (prefixStart > 0 && IDENTIFIER_CHAR.test(text[prefixStart - 1])) {
			prefixStart -= 1;
		}
		quoteStart = prefixStart;
	}
	const prefix = text.slice(prefixStart, offset);

	const qualifiers: string[] = [];
	// Walk back over `<identifier> .` pairs. Whitespace around the dots is allowed because it is
	// legal SQL, even if nobody writes it.
	let position = quoteStart;
	let more = true;
	while (more) {
		let scan = position;
		while (scan > 0 && (text[scan - 1] === ' ' || text[scan - 1] === '\t')) {
			scan -= 1;
		}
		if (scan <= 0 || text[scan - 1] !== '.') {
			break;
		}
		scan -= 1;
		while (scan > 0 && (text[scan - 1] === ' ' || text[scan - 1] === '\t')) {
			scan -= 1;
		}
		const identifier = readIdentifierBackwards(text, scan);
		if (identifier) {
			qualifiers.unshift(identifier.name);
			position = identifier.start;
		} else {
			more = false;
		}
	}

	const quoted = openQuote !== undefined;
	return {
		prefix,
		prefixStart,
		replaceStart: quoted ? prefixStart - 1 : prefixStart,
		qualifiers,
		suppressed: false,
		quoted,
	};
}

/**
 * The document with the identifier being typed swapped for a plain one, and where the cursor
 * moved to.
 *
 * A half typed quoted identifier (`o."Order T`) does not tokenize -- its opening quote swallows
 * the rest of the document -- which would cost the statement it is in its table completions, not
 * just the identifier. Replacing the partial identifier, opening quote included, makes it
 * ordinary SQL again.
 */
export function probeText(text: string, cursor: Cursor, offset: number): { text: string; offset: number } {
	const start = cursor.replaceStart;
	return {
		text: text.slice(0, start) + PLACEHOLDER + text.slice(offset),
		offset: start + PLACEHOLDER.length,
	};
}

File: extensions/positron-sql/sql-analyzer/src/cursor.rs
/*---------------------------------------------------------------------------------------------
 *  Copyright (C) 2026 Posit Software, PBC. All rights reserved.
 *  Licensed under the Elastic License 2.0. See LICENSE.txt for license information.
 *--------------------------------------------------------------------------------------------*/

//! What the statement under the cursor says the cursor can see, by parsing it with the hole filled
//! in.
//!
//! Completion is only ever asked about a statement halfway through being typed, and half typed SQL
//! does not parse: at `SELECT | FROM orders` there is no expression where one belongs, so
//! sqlparser returns an error and there is no tree at all -- no FROM clause to read, and nothing to
//! offer.
//!
//! So an identifier is written into the hole before the statement is parsed. `SELECT | FROM orders`
//! is parsed as `SELECT __positron_cursor__ FROM orders`, which does parse, and the tree it yields
//! is the tree of the statement the user is in the middle of writing. This is the trick IntelliJ
//! and rust-analyzer both use, for the same reason: an editor's parser is asked about incomplete
//! text far more often than about complete text, and a placeholder is much cheaper than a parser
//! that can recover.
//!
//! The marker earns its keep twice. It makes the statement parse, and -- because it comes back out
//! of the parser as a node with the cursor's own position -- it says *where in the tree* the cursor
//! ended up, which is the question [`crate::scopes`] can then answer properly. That is the whole
//! difference from [`crate::sources`], which reads the same tables off the tokens: a flat scan of
//! `WITH recent AS (SELECT id FROM orders) SELECT | FROM recent` finds `orders`, and offering its
//! columns is wrong, because the select the cursor is in cannot see them.
//!
//! When no attempt parses there is still [`crate::sources`] to fall back on, so the worst case is
//! the answer that was being given before.

use std::collections::HashSet;

use sqlparser::dialect::Dialect;
use sqlparser::keywords::Keyword;
use sqlparser::parser::Parser;
use sqlparser::tokenizer::{Span, Token, TokenWithSpan, Word};

use crate::diagnostics::RECURSION_LIMIT;
use crate::scopes::{self, Analysis};
use crate::sources::SourceRef;
use crate::statements::Statement;
use crate::text::TextIndex;

/// The identifier written into the hole the cursor is in.
///
/// Spelled so that it cannot collide with anything a person would type, since a real identifier of
/// the same name would be mistaken for the cursor. If one ever did, the cost is that completion is
/// scoped to the wrong part of a statement, not that anything breaks.
const MARKER: &str = "__positron_cursor__";

/// What the statement under the cursor can see from there.
pub struct AtCursor {
    /// The real tables in scope, nearest first. A table of an enclosing query is included: a
    /// correlated subquery may refer to one, so its columns belong in the list.
    pub sources: Vec<SourceRef>,
    /// The names the statement defines for itself and the cursor can refer to: CTEs, derived table
    /// aliases, and the names those are read under.
    pub locals: Vec<String>,
    /// Whether the cursor can see something whose columns cannot be known -- a CTE, a derived
    /// table, a table function. A column completed here may be one of those, so a caller that
    /// judges names must not judge them here.
    pub opaque: bool,
}

/// The tables and names in scope at an offset, or `None` if the statement could not be parsed even
/// with the hole filled in.
pub fn at(
    statement: &Statement,
    offset: u32,
    dialect: &dyn Dialect,
    index: &TextIndex,
) -> Option<AtCursor> {
    let (before, after) = split(&statement.tokens, offset, index);
    let marker = marker(offset, index);

    let whole: Vec<TokenWithSpan> =
        before.iter().chain(&marker).chain(after.iter()).cloned().collect();
    // Without the tail, for when the tail is what the statement cannot make sense of. Typing in
    // the middle of a statement usually leaves something ahead of the cursor that no longer fits
    // what is behind it, and the tables the cursor can see are nearly always behind it anyway.
    let prefix: Vec<TokenWithSpan> = before.iter().chain(&marker).cloned().collect();

    [whole, prefix].into_iter().find_map(|tokens| try_parse(close(tokens), dialect, index))
}

/// The marker as a token, positioned where the cursor is.
///
/// A real position rather than an empty span, because that is how it is found again: a node the
/// parser reports as coming from nowhere is skipped by [`crate::scopes`], which would leave the
/// marker in the tree and invisible in the analysis.
fn marker(offset: u32, index: &TextIndex) -> [TokenWithSpan; 1] {
    let at = index.location(offset);
    [TokenWithSpan {
        token: Token::Word(Word {
            value: MARKER.to_string(),
            quote_style: None,
            keyword: Keyword::NoKeyword,
        }),
        span: Span::new(at, at),
    }]
}

/// The tokens on either side of the cursor, with the word being typed dropped.
///
/// The token under the cursor is the word being completed, not a word the statement contains: at
/// `SELECT * FROM ord|` the user has typed three characters of a table name, and keeping them
/// would parse a statement about a table called `ord` rather than about the slot it sits in. The
/// marker takes its place.
fn split<'a>(
    tokens: &'a [TokenWithSpan],
    offset: u32,
    index: &TextIndex,
) -> (&'a [TokenWithSpan], &'a [TokenWithSpan]) {
    let mut cut = 0;
    let mut resume = tokens.len();

    for (at, token) in tokens.iter().enumerate() {
        let Some((start, end)) = index.span(token.span) else { continue };
        if end <= offset {
            cut = at + 1;
            continue;
        }
        // The first token the cursor has not run past. It either straddles the cursor, and is the
        // one being typed, or begins at or after it and is untouched.
        resume = if start < offset { at + 1 } else { at };
        break;
    }

    // A word ending exactly where the cursor is is also one being typed, since that is what the
    // cursor sits at the end of while it is being written.
    if let Some(last) = cut.checked_sub(1) {
        let typing = index
            .span(tokens[last].span)
            .map(|(start, end)| end == offset && start < offset)
            .unwrap_or(false);
        if typing && matches!(tokens[last].token, Token::Word(_)) {
            cut = last;
        }
    }

    (&tokens[..cut], &tokens[resume.max(cut)..])
}

/// Closes any parenthesis the cursor is still inside.
///
/// A subquery or a function call being typed is open by definition -- `SELECT * FROM (SELECT |`
/// -- and an unclosed one is the difference between a statement that parses and one that does
/// not, so the brackets the author has not reached yet are supplied.
fn close(mut tokens: Vec<TokenWithSpan>) -> Vec<TokenWithSpan> {
    let mut depth = 0i32;
    for token in &tokens {
        match token.token {
            Token::LParen => depth += 1,
            Token::RParen => depth -= 1,
            _ => {}
        }
    }
    for _ in 0..depth.max(0) {
        // No position: this is not in the document, and a node built out of it must not be
        // reported as though it were.
        tokens.push(TokenWithSpan { token: Token::RParen, span: Span::empty() });
    }
    tokens
}

fn try_parse(
    tokens: Vec<TokenWithSpan>,
    dialect: &dyn Dialect,
    index: &TextIndex,
) -> Option<AtCursor> {
    let parsed = Parser::new(dialect)
        .with_recursion_limit(RECURSION_LIMIT)
        .with_tokens_with_locations(tokens)
        .parse_statements()
        .ok()?;
    let analysis = scopes::analyze(&parsed, index);
    let scope = marker_scope(&analysis)?;
    Some(collect(&analysis, scope))
}

/// The scope the marker ended up in, or `None` if the parser dropped it.
///
/// Three places it can land, because three things can be written where a cursor sits: an
/// expression, the name of a table, and the alias of one -- `SELECT * FROM orders |` reads the
/// marker as what `orders` is being called.
fn marker_scope(analysis: &Analysis) -> Option<u32> {
    let column = analysis.columns.iter().find(|column| column.name == MARKER);
    if let Some(column) = column {
        return Some(column.scope);
    }
    analysis
        .tables
        .iter()
        .find(|table| table.name == MARKER || table.alias.as_deref() == Some(MARKER))
        .map(|table| table.scope)
}

/// Walks out from the cursor's scope, collecting what each level along the way can see.
fn collect(analysis: &Analysis, scope: u32) -> AtCursor {
    let mut found = AtCursor { sources: Vec::new(), locals: Vec::new(), opaque: false };
    // A select hands its tables up to the query around it, so walking out from a select reaches
    // the same reference twice; see `lift` in [`crate::scopes`].
    let mut seen: HashSet<u32> = HashSet::new();
    let mut named: HashSet<String> = HashSet::new();
    let mut at = Some(scope);

    // Scopes are numbered as they are opened, so a parent is always a lower number than its child
    // and the walk cannot loop. Bounded anyway: this runs in the extension host, where a hang is
    // the one failure the caller cannot recover from.
    for _ in 0..analysis.scopes.len() {
        let Some(id) = at else { break };
        let Some(entry) = analysis.scopes.get(id as usize) else { break };

        found.opaque |= entry.opaque;
        for local in &entry.locals {
            if local != MARKER && named.insert(local.to_lowercase()) {
                found.locals.push(local.clone());
            }
        }
        for at in &entry.tables {
            let Some(table) = analysis.tables.get(*at as usize) else { continue };
            if table.name == MARKER || !seen.insert(*at) {
                continue;
            }
            found.sources.push(SourceRef {
                name: table.name.clone(),
                schema: table.schema.clone(),
                catalog: table.catalog.clone(),
                // The marker is only ever read as an alias because the cursor sits where one would
                // go, which means the table has not been given one.
                alias: table.alias.clone().filter(|alias| alias != MARKER),
            });
        }

        at = entry.parent;
    }

    found
}
